backend/server/src/main.rs

```diff
@@ -79,8 +83,30 @@ async fn main() -> anyhow::Result<()> {
                 .to_string())
     );
 
-    tokio::spawn(async { unix_service.run().await });
-    web_service.run().await;
+    let cancellation_token = CancellationToken::new();
+
+    let eval_handle = tokio::spawn(eval_service.run(cancellation_token.clone()));
+    let unix_handle = tokio::spawn(unix_service.run(cancellation_token.clone()));
+    let web_handle = tokio::spawn(web_service.run(cancellation_token.clone()));
+
+    let mut sigterm = signal(SignalKind::terminate()).context("failed to get sigterm handle")?;
+    let mut sigint = signal(SignalKind::interrupt()).context("failed to get sigint handle")?;
+
+    tokio::select! {
```

Comment:
Could put these two in a `JoinSet` and then use `JoinSet::join_next`, if you want to remove the macro call. Personally, I am totally fine with leaving this as is, which might even be more efficient in this case, because it avoids "spawning" of the futures and they are instead polled "inline".

If you stay with the macro, you could make it `biased` to avoid the RNG call (the only reason you might want RNG is if you call `select!` in a loop and want to avoid starving some tasks, but that is not the case here, so we could remove it).

Reply:
i'll keep the macro and make it biased